# Electric Retail Service Territories

Exploratory data analysis for rural cooperatives and municipal utilities.

### Summary

The dataset has 2,949 records, 812 of which are rural cooperatives and 1,841 of which are municipal utilties. Relevant columns include the geometry, name (`NAME`), and state (`STATE`). The coordinate reference system is EPSG:4326.

### Exploration

In [1]:
import geopandas as gpd
import pandas as pd

In [2]:
pd.set_option("display.max_columns", None)

#### All Service Territories

In [3]:
gdf = gpd.read_file("../data/raw/bonus/retail/Electric_Retail_Service_Territories.zip")
gdf.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 2949 entries, 0 to 2948
Data columns (total 37 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   OBJECTID    2949 non-null   int64   
 1   ID          2949 non-null   object  
 2   NAME        2949 non-null   object  
 3   ADDRESS     2949 non-null   object  
 4   CITY        2949 non-null   object  
 5   STATE       2949 non-null   object  
 6   ZIP         2949 non-null   object  
 7   TELEPHONE   2949 non-null   object  
 8   TYPE        2949 non-null   object  
 9   COUNTRY     2949 non-null   object  
 10  NAICS_CODE  2949 non-null   object  
 11  NAICS_DESC  2949 non-null   object  
 12  SOURCE      2949 non-null   object  
 13  SOURCEDATE  2949 non-null   object  
 14  VAL_METHOD  2949 non-null   object  
 15  VAL_DATE    2949 non-null   object  
 16  WEBSITE     2949 non-null   object  
 17  REGULATED   2949 non-null   object  
 18  CNTRL_AREA  2949 non-null   object  
 19

In [11]:
final_matches = pd.read_excel("../reviewed_matches_final.xlsx", sheet_name="reviewed_matches")
final_matches.head(2)

MATCH_STRATEGY STATE  OBJECTID                    NAME_DHS  \
0  EXACT STATE, EXACT CITY    AK        29  KETCHIKAN PUBLIC UTILITIES   
1  EXACT STATE, FUZZY NAME    AK       117   CITY OF LARSEN BAY - (AK)   

                      NAME_CC  MANUALLY_MATCHED  MANUALLY_CORRECTED_CC_NAME  \
0  Ketchikan Public Utilities                 1                           0   
1          Larsen Bay Utility                 1                           0   

   SHOULD_DROP  
0            0  
1            0

In [23]:
corrected_municipal_utils = (final_matches
    .query("MANUALLY_MATCHED == 1")
    .loc[:, ["OBJECTID", "NAME_CC"]]
    .rename(columns={"NAME_CC": "NAME"})
    .sort_values(by="OBJECTID"))

corrected_municipal_utils.to_csv("../municipal_utility_name_matches.csv", index=False)

In [20]:
gdf.head(2)

OBJECTID     ID                    NAME           ADDRESS           CITY  \
0         1   1000  CITY OF AUGUSTA - (AR)     NOT AVAILABLE  NOT AVAILABLE   
1         2  10000            EVERGY METRO  1200 MAIN STREET    KANSAS CITY   

  STATE            ZIP      TELEPHONE            TYPE COUNTRY NAICS_CODE  \
0    AR  NOT AVAILABLE  NOT AVAILABLE       MUNICIPAL     USA       2211   
1    KS          64105  NOT AVAILABLE  INVESTOR OWNED     USA       2211   

                                          NAICS_DESC  \
0  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
1  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                              SOURCE  SOURCEDATE VAL_METHOD  \
0       EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
1  https://www.kcpl.com/-/media/indexedmedia/abou...  2021-10-06      OTHER   

     VAL_DATE                  WEBSITE      REGULATED  \
0  2018-10-12            NOT AVAILABLE  NOT AVAILABLE   
1  2021-04-30  https://www.evergy.com/  NOT AVAILABLE   

                                          CNTRL_AREA      PLAN_AREA  \
0  MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...  NOT AVAILABLE   
1                               SOUTHWEST POWER POOL  NOT AVAILABLE   

                     HOLDING_CO  SUMMR_PEAK  WINTR_PEAK  SUMMER_CAP  \
0        CITY OF AUGUSTA - (AR)   -999999.0   -999999.0   -999999.0   
1  KANSAS CITY POWER & LIGHT CO      3200.0      2484.0      4877.8   

   WINTER_CAP   NET_GEN  PURCHASED  NET_EX  RETAIL_MWH  WSALE_MWH  TOTAL_MWH  \
0   -999999.0     12823    -999999 -999999     -999999    -999999    -999999   
1   -999999.0  14911882    5624474       0    14223892    5957441   20536356   

   TRANS_MWH  CUSTOMERS  YEAR  Shape__Are  Shape__Len  \
0    -999999        916  2020    0.152271    2.625553   
1          0     562180  2020    0.525845   10.131485   

                                            geometry  
0  POLYGON ((-91.34713 35.43899, -91.34711 35.439...  
1  MULTIPOLYGON (((-94.66607 38.27027, -94.66941 ...

In [27]:
gdf["STATE"].sort_values().unique() # VI, AS, GU

array(['AB', 'AK', 'AL', 'AR', 'AS', 'AZ', 'BC', 'CA', 'CO', 'CT', 'DC',
       'DE', 'FL', 'GA', 'GU', 'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY',
       'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MP', 'MS', 'MT', 'NC',
       'ND', 'NE', 'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA',
       'PR', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VI', 'VT', 'WA',
       'WI', 'WV', 'WY'], dtype=object)

AB - Canada
AS - American Samoa
BC - Canada
DC - District of Columbia

GU - Guam
MP - Northern Mariana Islands
PR - Puerto Rico
VI - Virgin Islands


In [10]:
gdf.query("STATE == 'MP'")

OBJECTID     ID                              NAME        ADDRESS  \
2234      2235  57105  COMMONWEALTH UTILITY CORPORATION  NOT AVAILABLE   

               CITY STATE            ZIP      TELEPHONE   TYPE COUNTRY  \
2234  NOT AVAILABLE    MP  NOT AVAILABLE  NOT AVAILABLE  STATE     USA   

     NAICS_CODE                                         NAICS_DESC   SOURCE  \
2234       2211  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...  EIA 861   

      SOURCEDATE VAL_METHOD    VAL_DATE        WEBSITE      REGULATED  \
2234  2021-10-06      OTHER  2019-05-01  NOT AVAILABLE  NOT AVAILABLE   

         CNTRL_AREA      PLAN_AREA                        HOLDING_CO  \
2234  NOT AVAILABLE  NOT AVAILABLE  COMMONWEALTH UTILITY CORPORATION   

      SUMMR_PEAK  WINTR_PEAK  SUMMER_CAP  WINTER_CAP  NET_GEN  PURCHASED  \
2234        37.0        38.0   -999999.0   -999999.0   252882    -999999   

      NET_EX  RETAIL_MWH  WSALE_MWH  TOTAL_MWH  TRANS_MWH  CUSTOMERS  YEAR  \
2234       0      208299    -999999     252882          0      15541  2020   

      Shape__Are  Shape__Len  \
2234    0.169328    2.595992   

                                               geometry  
2234  MULTIPOLYGON (((145.27106 14.26232, 145.27659 ...

In [7]:
gdf.query("STATE == 'BC'")

OBJECTID    ID                            NAME  \
2793      2794   NA1                        FORTISBC   
2796      2797  NA12  SILVERSMITH POWER & LIGHT CORP   
2799      2800   NA2                        BC HYDRO   
2801      2802   NA3        HEMLOCK VALLEY UTILITIES   
2802      2803   NA4               KYUQUOT POWER LTD   

                                            ADDRESS        CITY STATE  \
2793               SUITE 100, 1975 SPRINGFIELD ROAD     KELOWNA    BC   
2796  SILVERSMITH POWER & LIGHT CORPORATION BOX 369  NEW DENVER    BC   
2799                                1155 MCGILL RD,    KAMLOOPS    BC   
2801                              HEMLOCK VALLEY RD     AGASSIZ    BC   
2802                                  NOT AVAILABLE   VANCOUVER    BC   

          ZIP       TELEPHONE            TYPE COUNTRY NAICS_CODE  \
2793  V1Y 7V7  (866) 436-7847  INVESTOR OWNED     CAN       2211   
2796  V0G 1S0  (250) 358-2247   NOT AVAILABLE     CAN       2211   
2799      V2C  (888) 769-3766         FEDERAL     CAN       2211   
2801  V0M 1A1  (604) 797-6882  INVESTOR OWNED     CAN       2211   
2802  V6G 2Z4  (604) 688-8271       MUNICIPAL     CAN       2211   

                                             NAICS_DESC  \
2793  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
2796  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
2799  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
2801  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
2802  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                                 SOURCE  SOURCEDATE  \
2793  http://www.atcoelectric.com/About-Us/Our-Servi...  2015-09-30   
2796                    http://canadiangis.com/data.php  2015-09-30   
2799  http://www.arcgis.com/home/item.html?id=dcbcdf...  2015-09-30   
2801         http://www.fortisbc.com/Pages/default.aspx  2015-09-30   
2802                              http://www.synex.com/  2015-09-30   

     VAL_METHOD    VAL_DATE  \
2793      OTHER  2020-05-18   
2796      OTHER  2015-09-30   
2799      OTHER  2015-09-30   
2801      OTHER  2015-09-30   
2802      OTHER  2015-09-30   

                                                WEBSITE      REGULATED  \
2793  https://outages.fortisbc.com/Outages/Home/Outa...  NOT AVAILABLE   
2796                            http://www.sandonbc-ca/  NOT AVAILABLE   
2799               https://www.bchydro.com/contact.html  NOT AVAILABLE   
2801                        hemlockvalleyutilities.com/  NOT AVAILABLE   
2802  http://www.electricpower-cc/company-kyuquot-po...  NOT AVAILABLE   

         CNTRL_AREA      PLAN_AREA                      HOLDING_CO  \
2793  NOT AVAILABLE  NOT AVAILABLE                        FORTISBC   
2796  NOT AVAILABLE  NOT AVAILABLE  SILVERSMITH POWER & LIGHT CORP   
2799  NOT AVAILABLE  NOT AVAILABLE                        BC HYDRO   
2801  NOT AVAILABLE  NOT AVAILABLE        HEMLOCK VALLEY UTILITIES   
2802  NOT AVAILABLE  NOT AVAILABLE               KYUQUOT POWER LTD   

      SUMMR_PEAK  WINTR_PEAK  SUMMER_CAP  WINTER_CAP  NET_GEN  PURCHASED  \
2793         0.0         0.0   -999999.0   -999999.0  -999999    -999999   
2796         0.0         0.0   -999999.0   -999999.0  -999999    -999999   
2799         0.0         0.0   -999999.0   -999999.0  -999999    -999999   
2801         0.0         0.0   -999999.0   -999999.0  -999999    -999999   
2802         0.0         0.0   -999999.0   -999999.0  -999999    -999999   

      NET_EX  RETAIL_MWH  WSALE_MWH  TOTAL_MWH  TRANS_MWH  CUSTOMERS  \
2793 -999999     -999999    -999999    -999999    -999999    -999999   
2796 -999999     -999999    -999999    -999999    -999999    -999999   
2799 -999999     -999999    -999999    -999999    -999999    -999999   
2801 -999999     -999999    -999999    -999999    -999999    -999999   
2802 -999999     -999999    -999999    -999999    -999999    -999999   

               YEAR  Shape__Are  Shape__Len  \
2793  NOT AVAILABLE    2.804753   10.264503   
27

In [11]:
import folium
map = folium.Map()
record = gdf.query("OBJECTID == 2235").iloc[0]
geom = record["geometry"]
print(record)
map.add_child(folium.GeoJson(geom))


OBJECTID                                                   2235
ID                                                        57105
NAME                           COMMONWEALTH UTILITY CORPORATION
ADDRESS                                           NOT AVAILABLE
CITY                                              NOT AVAILABLE
STATE                                                        MP
ZIP                                               NOT AVAILABLE
TELEPHONE                                         NOT AVAILABLE
TYPE                                                      STATE
COUNTRY                                                     USA
NAICS_CODE                                                 2211
NAICS_DESC    ELECTRIC POWER GENERATION, TRANSMISSION AND DI...
SOURCE                                                  EIA 861
SOURCEDATE                                           2021-10-06
VAL_METHOD                                                OTHER
VAL_DATE                                

In [4]:
gdf.columns

Index(['OBJECTID', 'ID', 'NAME', 'ADDRESS', 'CITY', 'STATE', 'ZIP',
       'TELEPHONE', 'TYPE', 'COUNTRY', 'NAICS_CODE', 'NAICS_DESC', 'SOURCE',
       'SOURCEDATE', 'VAL_METHOD', 'VAL_DATE', 'WEBSITE', 'REGULATED',
       'CNTRL_AREA', 'PLAN_AREA', 'HOLDING_CO', 'SUMMR_PEAK', 'WINTR_PEAK',
       'SUMMER_CAP', 'WINTER_CAP', 'NET_GEN', 'PURCHASED', 'NET_EX',
       'RETAIL_MWH', 'WSALE_MWH', 'TOTAL_MWH', 'TRANS_MWH', 'CUSTOMERS',
       'YEAR', 'Shape__Are', 'Shape__Len', 'geometry'],
      dtype='object')

In [5]:
gdf.head(2)

OBJECTID     ID                    NAME           ADDRESS           CITY  \
0         1   1000  CITY OF AUGUSTA - (AR)     NOT AVAILABLE  NOT AVAILABLE   
1         2  10000            EVERGY METRO  1200 MAIN STREET    KANSAS CITY   

  STATE            ZIP      TELEPHONE            TYPE COUNTRY NAICS_CODE  \
0    AR  NOT AVAILABLE  NOT AVAILABLE       MUNICIPAL     USA       2211   
1    KS          64105  NOT AVAILABLE  INVESTOR OWNED     USA       2211   

                                          NAICS_DESC  \
0  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
1  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                              SOURCE  SOURCEDATE VAL_METHOD  \
0       EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
1  https://www.kcpl.com/-/media/indexedmedia/abou...  2021-10-06      OTHER   

     VAL_DATE                  WEBSITE      REGULATED  \
0  2018-10-12            NOT AVAILABLE  NOT AVAILABLE   
1  2021-04-30  https://www.evergy.com/  NOT AVAILABLE   

                                          CNTRL_AREA      PLAN_AREA  \
0  MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...  NOT AVAILABLE   
1                               SOUTHWEST POWER POOL  NOT AVAILABLE   

                     HOLDING_CO  SUMMR_PEAK  WINTR_PEAK  SUMMER_CAP  \
0        CITY OF AUGUSTA - (AR)   -999999.0   -999999.0   -999999.0   
1  KANSAS CITY POWER & LIGHT CO      3200.0      2484.0      4877.8   

   WINTER_CAP   NET_GEN  PURCHASED  NET_EX  RETAIL_MWH  WSALE_MWH  TOTAL_MWH  \
0   -999999.0     12823    -999999 -999999     -999999    -999999    -999999   
1   -999999.0  14911882    5624474       0    14223892    5957441   20536356   

   TRANS_MWH  CUSTOMERS  YEAR  Shape__Are  Shape__Len  \
0    -999999        916  2020    0.152271    2.625553   
1          0     562180  2020    0.525845   10.131485   

                                            geometry  
0  POLYGON ((-91.34713 35.43899, -91.34711 35.439...  
1  MULTIPOLYGON (((-94.66607 38.27027, -94.66941 ...

In [6]:
gdf["TYPE"].sort_values().unique().tolist()

['COMMUNITY CHOICE AGGREGATOR',
 'COOPERATIVE',
 'FEDERAL',
 'INVESTOR OWNED',
 'MUNICIPAL',
 'MUNICIPAL MKTG AUTHORITY',
 'NOT AVAILABLE',
 'POLITICAL SUBDIVISION',
 'STATE',
 'WHOLESALE POWER MARKETER']

In [7]:
gdf.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

#### Municipalities

#### Load

In [8]:
municipal_types = ['MUNICIPAL', 'MUNICIPAL MKTG AUTHORITY']
municipalities = gdf.query("TYPE in @municipal_types")
municipalities.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Index: 1841 entries, 0 to 2948
Data columns (total 37 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   OBJECTID    1841 non-null   int64   
 1   ID          1841 non-null   object  
 2   NAME        1841 non-null   object  
 3   ADDRESS     1841 non-null   object  
 4   CITY        1841 non-null   object  
 5   STATE       1841 non-null   object  
 6   ZIP         1841 non-null   object  
 7   TELEPHONE   1841 non-null   object  
 8   TYPE        1841 non-null   object  
 9   COUNTRY     1841 non-null   object  
 10  NAICS_CODE  1841 non-null   object  
 11  NAICS_DESC  1841 non-null   object  
 12  SOURCE      1841 non-null   object  
 13  SOURCEDATE  1841 non-null   object  
 14  VAL_METHOD  1841 non-null   object  
 15  VAL_DATE    1841 non-null   object  
 16  WEBSITE     1841 non-null   object  
 17  REGULATED   1841 non-null   object  
 18  CNTRL_AREA  1841 non-null   object  
 19  PLA

In [9]:
municipalities.head(2)

OBJECTID    ID                    NAME        ADDRESS           CITY STATE  \
0         1  1000  CITY OF AUGUSTA - (AR)  NOT AVAILABLE  NOT AVAILABLE    AR   
6         7  1002  CITY OF AURELIA - (IA)  NOT AVAILABLE  NOT AVAILABLE    IA   

             ZIP      TELEPHONE       TYPE COUNTRY NAICS_CODE  \
0  NOT AVAILABLE  NOT AVAILABLE  MUNICIPAL     USA       2211   
6  NOT AVAILABLE  NOT AVAILABLE  MUNICIPAL     USA       2211   

                                          NAICS_DESC  \
0  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
6  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                         SOURCE  SOURCEDATE VAL_METHOD  \
0  EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
6  EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   

     VAL_DATE        WEBSITE      REGULATED  \
0  2018-10-12  NOT AVAILABLE  NOT AVAILABLE   
6  2018-10-12  NOT AVAILABLE  NOT AVAILABLE   

                                          CNTRL_AREA      PLAN_AREA  \
0  MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...  NOT AVAILABLE   
6  MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...  NOT AVAILABLE   

               HOLDING_CO  SUMMR_PEAK  WINTR_PEAK  SUMMER_CAP  WINTER_CAP  \
0  CITY OF AUGUSTA - (AR)   -999999.0   -999999.0   -999999.0   -999999.0   
6  CITY OF AURELIA - (IA)   -999999.0   -999999.0   -999999.0   -999999.0   

   NET_GEN  PURCHASED  NET_EX  RETAIL_MWH  WSALE_MWH  TOTAL_MWH  TRANS_MWH  \
0    12823    -999999 -999999     -999999    -999999    -999999    -999999   
6     7441    -999999 -999999     -999999    -999999    -999999    -999999   

   CUSTOMERS  YEAR  Shape__Are  Shape__Len  \
0        916  2020    0.152271    2.625553   
6        518  2020    0.164284    1.639697   

                                            geometry  
0  POLYGON ((-91.34713 35.43899, -91.34711 35.439...  
6  POLYGON ((-95.85936 42.90908, -95.85125 42.909...

#### Clean Custom

In [11]:
updated_names = pd.read_csv(
    filepath_or_buffer="../data/raw/bonus/retail/utility name - city name - Sheet1.csv",
    names=["NAME", "CITY"])
updated_names.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1948 entries, 0 to 1947
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   NAME    1948 non-null   object
 1   CITY    1948 non-null   object
dtypes: object(2)
memory usage: 30.6+ KB


In [12]:
updated_names.head(2)

NAME          CITY
0  Ak-Chin Energy Services  Maricopa, AZ
1             City of Mesa      Mesa, AZ

In [13]:
# Split into city and state columns
def split_addr(city: str):
    if city == ",":
        return None, None
    items = city.upper().split(", ")
    return items[0], items[1]

updated_names["CITY"], updated_names["STATE"] = zip(*updated_names["CITY"].apply(split_addr))

In [14]:
updated_names.head(2)

NAME      CITY STATE
0  Ak-Chin Energy Services  MARICOPA    AZ
1             City of Mesa      MESA    AZ

In [15]:
updated_names["NAME"] = updated_names["NAME"].str.upper()
updated_names.head(2)

NAME      CITY STATE
0  AK-CHIN ENERGY SERVICES  MARICOPA    AZ
1             CITY OF MESA      MESA    AZ

In [16]:
updated_names = updated_names.sort_values(by="NAME").replace({",": None})
updated_names.head(100)

NAME        CITY STATE
688          ABBESVILLE PUBLIC UTILITIES   ABBEVILLE    SC
1289        ABERDEEN ELECTRIC DEPARTMENT    ABERDEEN    MS
1863                        ADA TOWNSHIP        None  None
345         ADA WATER & LIGHT DEPARTMENT         ADA    MN
346   ADRIAN PUBLIC UTILITIES COMMISSION      ADRIAN    MN
...                                  ...         ...   ...
671                  BOROUGH OF OLYPHANT    OLYPHANT    PA
630                 BOROUGH OF PEMBERTON   PEMBERTON    NJ
672                  BOROUGH OF PERKASIE    PERKASIE    PA
673                BOROUGH OF QUAKERTOWN  QUAKERTOWN    PA
674                  BOROUGH OF ROYALTON    ROYALTON    PA

[100 rows x 3 columns]

#### Clean Export

In [17]:
municipalities["NAME"] = municipalities["NAME"].str.replace("COMM", "COMMISSION")
municipalities["NAME"] = municipalities["NAME"].apply(lambda n: n.split("-")[0].strip())
municipalities.head(2)

/home/launagreer/clinic/climate-cabinet-tax-credit-map/env/lib/python3.8/site-packages/geopandas/geodataframe.py:1538: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)
/home/launagreer/clinic/climate-cabinet-tax-credit-map/env/lib/python3.8/site-packages/geopandas/geodataframe.py:1538: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)


OBJECTID    ID             NAME        ADDRESS           CITY STATE  \
0         1  1000  CITY OF AUGUSTA  NOT AVAILABLE  NOT AVAILABLE    AR   
6         7  1002  CITY OF AURELIA  NOT AVAILABLE  NOT AVAILABLE    IA   

             ZIP      TELEPHONE       TYPE COUNTRY NAICS_CODE  \
0  NOT AVAILABLE  NOT AVAILABLE  MUNICIPAL     USA       2211   
6  NOT AVAILABLE  NOT AVAILABLE  MUNICIPAL     USA       2211   

                                          NAICS_DESC  \
0  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
6  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                         SOURCE  SOURCEDATE VAL_METHOD  \
0  EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
6  EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   

     VAL_DATE        WEBSITE      REGULATED  \
0  2018-10-12  NOT AVAILABLE  NOT AVAILABLE   
6  2018-10-12  NOT AVAILABLE  NOT AVAILABLE   

                                          CNTRL_AREA      PLAN_AREA  \
0  MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...  NOT AVAILABLE   
6  MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...  NOT AVAILABLE   

               HOLDING_CO  SUMMR_PEAK  WINTR_PEAK  SUMMER_CAP  WINTER_CAP  \
0  CITY OF AUGUSTA - (AR)   -999999.0   -999999.0   -999999.0   -999999.0   
6  CITY OF AURELIA - (IA)   -999999.0   -999999.0   -999999.0   -999999.0   

   NET_GEN  PURCHASED  NET_EX  RETAIL_MWH  WSALE_MWH  TOTAL_MWH  TRANS_MWH  \
0    12823    -999999 -999999     -999999    -999999    -999999    -999999   
6     7441    -999999 -999999     -999999    -999999    -999999    -999999   

   CUSTOMERS  YEAR  Shape__Are  Shape__Len  \
0        916  2020    0.152271    2.625553   
6        518  2020    0.164284    1.639697   

                                            geometry  
0  POLYGON ((-91.34713 35.43899, -91.34711 35.439...  
6  POLYGON ((-95.85936 42.90908, -95.85125 42.909...

In [18]:
municipalities[["NAME", "CITY", "STATE"]].sort_values(by="NAME").head(100)

NAME           CITY STATE
778      ADRIAN PUBLIC UTILITIES COMMISSION         ADRIAN    MN
1110     AITKIN PUBLIC UTILITIES COMMISSION  NOT AVAILABLE    MN
1594                ALAMEDA MUNICIPAL POWER  NOT AVAILABLE    CA
2825                   ALBANY UTILITY BOARD  NOT AVAILABLE    GA
1715  ALBERTVILLE MUNICIPAL UTILITIES BOARD    ALBERTVILLE    AL
...                                     ...            ...   ...
1697                         CITY OF ALBANY         ALBANY    MO
1704                         CITY OF ALBANY  NOT AVAILABLE    IL
1707                      CITY OF ALBEMARLE  NOT AVAILABLE    NC
1721                         CITY OF ALBION         ALBION    ID
2249                CITY OF ALCOA UTILITIES  NOT AVAILABLE    TN

[100 rows x 3 columns]

In [19]:
municipalities = municipalities.replace({"NOT AVAILABLE": None})
municipalities.head(100)

OBJECTID     ID                  NAME ADDRESS  CITY STATE   ZIP  \
0           1   1000       CITY OF AUGUSTA    None  None    AR  None   
6           7   1002       CITY OF AURELIA    None  None    IA  None   
7           8  10025        CITY OF KAPLAN    None  None    LA  None   
8           9  10037        CITY OF KASOTA    None  None    MN  None   
9          10   1004        CITY OF AURORA    None  None    SD  None   
..        ...    ...                   ...     ...   ...   ...   ...   
139       140  10881        CITY OF LEHIGH    None  None    IA  None   
140       141  10882  BOROUGH OF LEHIGHTON    None  None    PA  None   
141       142  10886      VILLAGE OF LEIGH    None  None    NE  None   
142       143  10892        CITY OF LELAND    None  None    MS  None   
143       144  10906        CITY OF LENOIR    None  None    TN  None   

    TELEPHONE       TYPE COUNTRY NAICS_CODE  \
0        None  MUNICIPAL     USA       2211   
6        None  MUNICIPAL     USA       2211   
7        None  MUNICIPAL     USA       2211   
8        None  MUNICIPAL     USA       2211   
9        None  MUNICIPAL     USA       2211   
..        ...        ...     ...        ...   
139      None  MUNICIPAL     USA       2211   
140      None  MUNICIPAL     USA       2211   
141      None  MUNICIPAL     USA       2211   
142      None  MUNICIPAL     USA       2211   
143      None  MUNICIPAL     USA       2211   

                                            NAICS_DESC  \
0    ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
6    ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
7    ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
8    ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
9    ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
..                                                 ...   
139  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
140  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
141  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
142  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
143  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                                SOURCE  SOURCEDATE VAL_METHOD  \
0         EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
6         EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
7    EIA 861, Urban Areas - Cartographic Boundary S...  2021-10-06      OTHER   
8    EIA 861, https://gisdata.mn.gov/dataset/util-eusa  2021-10-06      OTHER   
9    EIA 861, Urban Areas - Cartographic Boundary S...  2021-10-06      OTHER   
..                                                 ...         ...        ...   
139       EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
140       EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
141       EIA 861, TIGER/Line Shapefiles - U.S. Census  2019-10-01      OTHER   
142  EIA 861, Urban Areas - Cartographic Boundary S...  2021-10-06      OTHER   
143  EIA 861, https://www.tva.gov/file_source/TVA/S...  2021-10-06      OTHER   

       VAL_DATE WEBSITE  REGULATED  \
0    2018-10-12    None       None   
6    2018-10-12    None       None   
7    2018-10-12    None       None   
8    2018-10-12    None       None   
9    2018-10-12    None       None   
..          ...     ...        ...   
139  2018-10-12    None       None   
140  2018-10-12    None       None   
141  2018-10-12    None       None   
142  2018-10-12    None       None   
143  2018-10-12    None  REGULATED   

                                            CNTRL_AREA PLAN_AREA  \
0    MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...      None   
6    MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...      None   
7    MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...      None   
8    MIDCONTINENT INDEPENDENT TRANSMISSION SYSTEM O...      None   
9                                 SOUTHWEST POWER POOL      None   
..                                      

#### Match - Exact Name, Exact State

Attempt to match on common name and state

In [20]:
municipalities["IS_EXPORT"] = 1
updated_names["IS_CUSTOM"] = 1
combined = municipalities[["OBJECTID", "NAME", "STATE", "CITY", "IS_EXPORT", ]].merge(
    right=updated_names,
    how="outer",
    on=["NAME", "STATE"],
    suffixes=["_EXPORT", "_CUSTOM"]
)
combined = combined[[
    "OBJECTID",
    "NAME",
    "STATE",
    "CITY_EXPORT",
    "CITY_CUSTOM",
    "IS_EXPORT",
    "IS_CUSTOM"
]]
combined.to_csv("muni_matching.csv", index=False)

In [21]:
print(len(combined))

2664


In [22]:
agreements = combined.query("(IS_EXPORT == 1) and (IS_CUSTOM == 1)")
print(len(agreements))

1125


In [23]:
agreements

OBJECTID                            NAME STATE CITY_EXPORT  \
0          1.0                 CITY OF AUGUSTA    AR        None   
1          7.0                 CITY OF AURELIA    IA        None   
2          8.0                  CITY OF KAPLAN    LA        None   
3          9.0                  CITY OF KASOTA    MN        None   
4         10.0                  CITY OF AURORA    SD        None   
...        ...                             ...   ...         ...   
1832    2941.0            CITY OF NATCHITOCHES    LA        None   
1833    2942.0     CLARKSDALE PUBLIC UTILITIES    MS  CLARKSDALE   
1834    2943.0                CITY OF SULLIVAN    MO        None   
1836    2945.0     SHAWANO MUNICIPAL UTILITIES    WI        None   
1838    2947.0  WAGONER PUBLIC WORKS AUTHORITY    OK        None   

       CITY_CUSTOM  IS_EXPORT  IS_CUSTOM  
0          AUGUSTA        1.0        1.0  
1          AURELIA        1.0        1.0  
2           KAPLAN        1.0        1.0  
3           KASOTA        1.0        1.0  
4           AURORA        1.0        1.0  
...            ...        ...        ...  
1832  NATCHITOCHES        1.0        1.0  
1833    CLARKSDALE        1.0        1.0  
1834      SULLIVAN        1.0        1.0  
1836       SHAWANO        1.0        1.0  
1838       WAGONER        1.0        1.0  

[1125 rows x 7 columns]

There are 1,125 records that match on name and state alone. Initialize list of corrected names.

In [24]:
corrected = agreements[["OBJECTID", "NAME"]]
corrected["OBJECTID"] = corrected["OBJECTID"].astype(int)
corrected

/tmp/ipykernel_17773/2083669106.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  corrected["OBJECTID"] = corrected["OBJECTID"].astype(int)


OBJECTID                            NAME
0            1                 CITY OF AUGUSTA
1            7                 CITY OF AURELIA
2            8                  CITY OF KAPLAN
3            9                  CITY OF KASOTA
4           10                  CITY OF AURORA
...        ...                             ...
1832      2941            CITY OF NATCHITOCHES
1833      2942     CLARKSDALE PUBLIC UTILITIES
1834      2943                CITY OF SULLIVAN
1836      2945     SHAWANO MUNICIPAL UTILITIES
1838      2947  WAGONER PUBLIC WORKS AUTHORITY

[1125 rows x 2 columns]

#### Match - Exact City, Exact State

Count the number of remaining records without matches:

In [25]:
disagreements = combined.query("(IS_EXPORT != IS_EXPORT) or (IS_CUSTOM != IS_CUSTOM)")
print(len(disagreements))

1539


The remaining 1,539 do not match. Attempt to match on city and state next.

In [26]:
custom = disagreements.query("IS_CUSTOM == 1")[["OBJECTID", "NAME", "STATE", "CITY_CUSTOM"]]
export = disagreements.query("IS_EXPORT == 1")[["OBJECTID", "NAME", "STATE", "CITY_EXPORT"]]
print(len(custom))
print(len(export))

823
716


In [27]:
col_map = {
    "OBJECTID_x": "OBJECTID_CUSTOM",
    "OBJECTID_y": "OBJECTID_EXPORT",
    "NAME_x": "NAME_CUSTOM",
    "NAME_y": "NAME_EXPORT",
    "CITY_EXPORT": "CITY_EXPORT",
    "CITY_CUSTOM": "CITY_CUSTOM",
    "STATE": "STATE"
}
combined2 = custom.merge(
    right=export,
    how="left",
    left_on=["STATE", "CITY_CUSTOM"],
    right_on=["STATE", "CITY_EXPORT"]
)[list(col_map.keys())].rename(columns=col_map)
combined2.head(5)

OBJECTID_CUSTOM  OBJECTID_EXPORT                      NAME_CUSTOM  \
0              NaN           1888.0      ABBESVILLE PUBLIC UTILITIES   
1              NaN              NaN     ABERDEEN ELECTRIC DEPARTMENT   
2              NaN              NaN                     ADA TOWNSHIP   
3              NaN              NaN     ADA WATER & LIGHT DEPARTMENT   
4              NaN              NaN  ADVANCE MUNICIPAL LIGHT & POWER   

         NAME_EXPORT CITY_EXPORT CITY_CUSTOM STATE  
0  CITY OF ABBEVILLE   ABBEVILLE   ABBEVILLE    SC  
1                NaN         NaN    ABERDEEN    MS  
2                NaN         NaN        None  None  
3                NaN         NaN         ADA    MN  
4                NaN         NaN     ADVANCE    IN

In [28]:
combined2.query("OBJECTID_EXPORT == OBJECTID_EXPORT")

OBJECTID_CUSTOM  OBJECTID_EXPORT                           NAME_CUSTOM  \
0                NaN           1888.0           ABBESVILLE PUBLIC UTILITIES   
8                NaN           1722.0            ALBION LIGHT & WATER PLANT   
10               NaN           1835.0            ALGONA MUNICIPAL UTILITIES   
13               NaN           1810.0                         ALP UTILITIES   
16               NaN           2275.0              ANAHEIM PUBLIC UTILITIES   
..               ...              ...                                   ...   
817              NaN           1642.0          WORTHINGTON PUBLIC UTILITIES   
818              NaN           1643.0      WRANGELL MUNICIPAL LIGHT & POWER   
819              NaN           1645.0          WYANDOTTE MUNICIPAL SERVICES   
821              NaN           1651.0  YAZOO CITY PUBLIC SERVICE COMMISSION   
822              NaN           1657.0         ZEELAND BOARD OF PUBLIC WORKS   

                              NAME_EXPORT  CITY_EXPORT  CITY_CUSTOM STATE  
0                       CITY OF ABBEVILLE    ABBEVILLE    ABBEVILLE    SC  
8                          CITY OF ALBION       ALBION       ALBION    ID  
10                         CITY OF ALGONA       ALGONA       ALGONA    IA  
13                     CITY OF ALEXANDRIA   ALEXANDRIA   ALEXANDRIA    MN  
16                        CITY OF ANAHEIM      ANAHEIM      ANAHEIM    CA  
..                                    ...          ...          ...   ...  
817                   CITY OF WORTHINGTON  WORTHINGTON  WORTHINGTON    MN  
818                      CITY OF WRANGELL     WRANGELL     WRANGELL    AK  
819   WYANDOTTE MUNICIPAL SERV COMMISSION    WYANDOTTE    WYANDOTTE    MI  
821  PUBLIC SERV COMMISSION OF YAZOO CITY   YAZOO CITY   YAZOO CITY    MS  
822                       CITY OF ZEELAND      ZEELAND      ZEELAND    MI  

[262 rows x 7 columns]

Of the remaining records, 262 match on city and state.

In [29]:
corrected2 = combined2.query("OBJECTID_EXPORT == OBJECTID_EXPORT")[["OBJECTID_EXPORT", "NAME_CUSTOM"]]
corrected2 = corrected2.rename(columns={"OBJECTID_EXPORT": "OBJECTID", "NAME_CUSTOM": "NAME"})
corrected2["OBJECTID"] = corrected2["OBJECTID"].astype(int)
corrected2.head()

OBJECTID                         NAME
0       1888  ABBESVILLE PUBLIC UTILITIES
8       1722   ALBION LIGHT & WATER PLANT
10      1835   ALGONA MUNICIPAL UTILITIES
13      1810                ALP UTILITIES
16      2275     ANAHEIM PUBLIC UTILITIES

Update DataFrame of corrected names

In [30]:
corrected = pd.concat([corrected, corrected2]).sort_values(by="OBJECTID")
corrected

OBJECTID                            NAME
0            1                 CITY OF AUGUSTA
1            7                 CITY OF AURELIA
2            8                  CITY OF KAPLAN
3            9                  CITY OF KASOTA
4           10                  CITY OF AURORA
...        ...                             ...
1832      2941            CITY OF NATCHITOCHES
1833      2942     CLARKSDALE PUBLIC UTILITIES
1834      2943                CITY OF SULLIVAN
1836      2945     SHAWANO MUNICIPAL UTILITIES
1838      2947  WAGONER PUBLIC WORKS AUTHORITY

[1387 rows x 2 columns]

#### Match - Fuzzy Name, Exact State

Examine remaining records that don't match on name and state or city and state but are present in the custom dataset.

In [31]:
custom_only = combined2.query("OBJECTID_EXPORT != OBJECTID_EXPORT").sort_values(by="NAME_CUSTOM")
custom_only

OBJECTID_CUSTOM  OBJECTID_EXPORT  \
1                NaN              NaN   
2                NaN              NaN   
3                NaN              NaN   
4                NaN              NaN   
5                NaN              NaN   
..               ...              ...   
812              NaN              NaN   
813              NaN              NaN   
814              NaN              NaN   
815              NaN              NaN   
820              NaN              NaN   

                                           NAME_CUSTOM NAME_EXPORT  \
1                         ABERDEEN ELECTRIC DEPARTMENT         NaN   
2                                         ADA TOWNSHIP         NaN   
3                         ADA WATER & LIGHT DEPARTMENT         NaN   
4                      ADVANCE MUNICIPAL LIGHT & POWER         NaN   
5                              AK-CHIN ENERGY SERVICES         NaN   
..                                                 ...         ...   
812              WISCONSIN DELLS WATER & LIGHT UTILITY         NaN   
813  WISCONSIN RAPIDS WATER WORKS & LIGHTING COMMIS...         NaN   
814         WONEWOC MUNICIPAL WATER & LIGHT DEPARTMENT         NaN   
815                       WOODBINE MUNICIPAL UTILITIES         NaN   
820                 WYNNEWOOD CITY UTILITIES AUTHORITY         NaN   

    CITY_EXPORT       CITY_CUSTOM STATE  
1           NaN          ABERDEEN    MS  
2           NaN              None  None  
3           NaN               ADA    MN  
4           NaN           ADVANCE    IN  
5           NaN          MARICOPA    AZ  
..          ...               ...   ...  
812         NaN   WISCONSIN DELLS    WI  
813         NaN  WISCONSIN RAPIDS    WI  
814         NaN           WONEWOC    WI  
815         NaN          WOODBINE    IA  
820         NaN         WYNNEWOOD    OK  

[561 rows x 7 columns]

There are 561 rows in the custom dataset that still lack a match.

In [32]:
col_map = {
    "OBJECTID_x": "OBJECTID_EXPORT",
    "OBJECTID_y": "OBJECTID_CUSTOM",
    "NAME_x": "NAME_EXPORT",
    "NAME_y": "NAME_CUSTOM",
    "CITY_EXPORT": "CITY_EXPORT",
    "CITY_CUSTOM": "CITY_CUSTOM",
    "STATE": "STATE"
}
combined3 = export.merge(
    right=custom,
    how="left",
    left_on=["STATE", "CITY_EXPORT"],
    right_on=["STATE", "CITY_CUSTOM"]
)[list(col_map.keys())].rename(columns=col_map)
combined3.head(5)

OBJECTID_EXPORT  OBJECTID_CUSTOM                   NAME_EXPORT  \
0             12.0              NaN              CITY OF KAUKAUNA   
1             18.0              NaN     KEEWATIN PUBLIC UTILITIES   
2             19.0              NaN                CITY OF AUSTIN   
3             20.0              NaN  KENNEBUNK LIGHT & POWER DIST   
4             22.0              NaN               CITY OF KENNETT   

                       NAME_CUSTOM CITY_EXPORT CITY_CUSTOM STATE  
0               KAUKAUNA UTILITIES    KAUKAUNA    KAUKAUNA    WI  
1                              NaN        None         NaN    MN  
2                 AUSTIN UTILITIES      AUSTIN      AUSTIN    MN  
3                              NaN        None         NaN    ME  
4  KENNETT CITY LIGHT, GAS & WATER     KENNETT     KENNETT    MO

In [33]:
combined3.query("NAME_CUSTOM == NAME_CUSTOM").sort_values(by="NAME_EXPORT")

OBJECTID_EXPORT  OBJECTID_CUSTOM  \
73             279.0              NaN   
320           1318.0              NaN   
126            471.0              NaN   
428           1811.0              NaN   
695           2894.0              NaN   
..               ...              ...   
139            517.0              NaN   
350           1459.0              NaN   
366           1520.0              NaN   
708           2920.0              NaN   
395           1645.0              NaN   

                                        NAME_EXPORT  \
73                     BANCROFT MUNICIPAL UTILITIES   
320                        BLUE EARTH LIGHT & WATER   
126  BOARD OF WATER ELECTRIC & COMMISSIONUNICATIONS   
428                             CANBY UTILITY BOARD   
695                  CARROLLTON BOARD OF PUBLIC WKS   
..                                              ...   
139                        VILLAGE OF NEW KNOXVILLE   
350                                VILLAGE OF VIOLA   
366                  WAVERLY MUNICIPAL ELEC UTILITY   
708                  WEATHERFORD MUN UTILITY SYSTEM   
395             WYANDOTTE MUNICIPAL SERV COMMISSION   

                              NAME_CUSTOM    CITY_EXPORT    CITY_CUSTOM STATE  
73             BANCROFT MUNICIPAL UTILITY       BANCROFT       BANCROFT    IA  
320   BLUE EARTH LIGHT & WATER DEPARTMENT     BLUE EARTH     BLUE EARTH    MN  
126               MUSCATINE POWER & WATER      MUSCATINE      MUSCATINE    IA  
428                         CANBY UTILITY          CANBY          CANBY    OR  
695                    CITY OF CARROLLTON     CARROLLTON     CARROLLTON    MO  
..                                    ...            ...            ...   ...  
139      NEW KNOXVILLE MUNICIPAL ELECTRIC  NEW KNOXVILLE  NEW KNOXVILLE    OH  
350      VIOLA MUNICIPAL ELECTRIC UTILITY          VIOLA          VIOLA    WI  
366                     WAVERLY UTILITIES        WAVERLY        WAVERLY    IA  
708  WEATHERFORD MUNICIPAL UTILITY SYSTEM    WEATHERFORD    WEATHERFORD    TX  
395          WYANDOTTE MUNICIPAL SERVICES      WYANDOTTE      WYANDOTTE    MI  

[262 rows x 7 columns]

In [34]:
export_only = combined3.query("NAME_CUSTOM != NAME_CUSTOM").sort_values(by="NAME_EXPORT")
export_only

OBJECTID_EXPORT  OBJECTID_CUSTOM                          NAME_EXPORT  \
657           2826.0              NaN                 ALBANY UTILITY BOARD   
39             160.0              NaN   BAGLEY PUBLIC UTILITIES COMMISSION   
119            444.0              NaN                  BARTON VILLAGE, INC   
137            511.0              NaN            BATH ELECTRIC GAS & WATER   
218            899.0              NaN                        BENTON COUNTY   
..               ...              ...                                  ...   
367           1528.0              NaN          WEAKLEY COUNTY MUN ELEC SYS   
385           1614.0              NaN        WISCONSIN DELLS ELECTRIC UTIL   
388           1622.0              NaN  WISCONSIN RAPIDS W W & L COMMISSION   
389           1628.0              NaN        WONEWOC ELECTRIC & WATER UTIL   
386           1619.0              NaN                          WPPI ENERGY   

    NAME_CUSTOM  CITY_EXPORT CITY_CUSTOM STATE  
657         NaN         None         NaN    GA  
39          NaN         None         NaN    MN  
119         NaN       BARTON         NaN    VT  
137         NaN         None         NaN    NY  
218         NaN         None         NaN    TN  
..          ...          ...         ...   ...  
367         NaN         None         NaN    TN  
385         NaN         None         NaN    WI  
388         NaN         None         NaN    WI  
389         NaN         None         NaN    WI  
386         NaN  SUN PRAIRIE         NaN    WI  

[456 rows x 7 columns]

There are 456 rows in the exported dataset that still lack a match.

In [35]:

state_abbrev_map = {
    'AK': 'Alaska',
    'AL': 'Alabama',
    'AR': 'Arkansas',
    'AZ': 'Arizona',
    'CA': 'California',
    'CO': 'Colorado',
    'CT': 'Connecticut',
    'DC': 'District of Columbia',
    'DE': 'Delaware',
    'FL': 'Florida',
    'GA': 'Georgia',
    'HI': 'Hawaii',
    'IA': 'Iowa',
    'ID': 'Idaho',
    'IL': 'Illinois',
    'IN': 'Indiana',
    'KS': 'Kansas',
    'KY': 'Kentucky',
    'LA': 'Louisiana',
    'MA': 'Massachusetts',
    'MD': 'Maryland',
    'ME': 'Maine',
    'MI': 'Michigan',
    'MN': 'Minnesota',
    'MO': 'Missouri',
    'MS': 'Mississippi',
    'MT': 'Montana',
    'NC': 'North Carolina',
    'ND': 'North Dakota',
    'NE': 'Nebraska',
    'NH': 'New Hampshire',
    'NJ': 'New Jersey',
    'NM': 'New Mexico',
    'NV': 'Nevada',
    'NY': 'New York',
    'OH': 'Ohio',
    'OK': 'Oklahoma',
    'OR': 'Oregon',
    'PA': 'Pennsylvania',
    'RI': 'Rhode Island',
    'SC': 'South Carolina',
    'SD': 'South Dakota',
    'TN': 'Tennessee',
    'TX': 'Texas',
    'UT': 'Utah',
    'VA': 'Virginia',
    'VT': 'Vermont',
    'WA': 'Washington',
    'WI': 'Wisconsin',
    'WV': 'West Virginia',
    'WY': 'Wyoming',
    'PR': 'Puerto Rico'
}

In [36]:
from fuzzywuzzy import fuzz

/home/launagreer/clinic/climate-cabinet-tax-credit-map/env/lib/python3.8/site-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


In [37]:
def clean_name(name: str):
    excluded = [
        "BOARD",
        "CITY",
        "TOWN",
        "VILLAGE",
        "DEPT.",
        "DEPARTMENT",
        "OF",
        "ELECTRICITY",
        "ELECTRIC",
        "ELEC",
        "CO.",
        "COMPANY",
        "UTILITIES",
        "UTILITY",
        "UTIL",
        "POWER",
        "PLANT",
        "SYSTEMS",
        "COMMISSION",
        "COMM"
    ]
    for term in excluded:
        name = name.replace(term, "")
    return name.strip()

def calculate_similarity(row: pd.Series):
    name1 = clean_name(row["NAME_EXPORT"])
    name2 = clean_name(row["NAME_CUSTOM"])
    return fuzz.token_sort_ratio(name1, name2)

In [38]:
def process_group(g):
    df = pd.DataFrame(g)
    top_results = df[df["ratio"] == df["ratio"].max()]
    return top_results

In [39]:
export_only.head()

OBJECTID_EXPORT  OBJECTID_CUSTOM                         NAME_EXPORT  \
657           2826.0              NaN                ALBANY UTILITY BOARD   
39             160.0              NaN  BAGLEY PUBLIC UTILITIES COMMISSION   
119            444.0              NaN                 BARTON VILLAGE, INC   
137            511.0              NaN           BATH ELECTRIC GAS & WATER   
218            899.0              NaN                       BENTON COUNTY   

    NAME_CUSTOM CITY_EXPORT CITY_CUSTOM STATE  
657         NaN        None         NaN    GA  
39          NaN        None         NaN    MN  
119         NaN      BARTON         NaN    VT  
137         NaN        None         NaN    NY  
218         NaN        None         NaN    TN

In [40]:
top_matches_export = None
top_matches_custom = None

for state in list(state_abbrev_map.keys()):
    # Subsest records by state
    export_subset = export_only.query("STATE == @state")
    custom_subset = custom_only.query("STATE == @state")
    print(f"{len(export_subset)} record(s) in export, {len(custom_subset)} in custom")

    # Skip processing if either DataFrame lacks reords
    # (i.e., no matches are possible)
    if not len(export_subset) or not len(custom_subset):
        print(f"Missing records in both dataframes for state \"{state}\". Skipping.")
        continue

    # Otherwise, perform cross join of names
    cross_join = export_subset[["OBJECTID_EXPORT", "NAME_EXPORT"]].merge(
        right=custom_subset[["NAME_CUSTOM"]],
        how="cross")
    
    # Calculate similarity score (here, fuzzywuzzy's token ratio)
    cross_join["ratio"] = cross_join.apply(calculate_similarity, axis=1)

    # Calculate best match for each name in export dataset
    max_scores_export = (cross_join[["NAME_EXPORT", "NAME_CUSTOM", "ratio"]]
            .groupby(by=["NAME_EXPORT"])
            .apply(process_group)
            .rename(columns={"NAME_EXPORT": "name_export"})
            .reset_index()
            .drop(columns=["NAME_EXPORT", "level_1"])
            .rename(columns={"name_export": "NAME_EXPORT", "ratio": "RATIO"}))
    max_scores_export["STATE"] = state
    max_scores_export = max_scores_export.merge(
        right=export_only[["OBJECTID_EXPORT", "STATE", "NAME_EXPORT"]],
        how="left",
        on=["STATE", "NAME_EXPORT"]
    )
    max_scores_export["OBJECTID_EXPORT"] = max_scores_export["OBJECTID_EXPORT"].astype(int)
    max_scores_export = max_scores_export[["OBJECTID_EXPORT", "STATE", "NAME_EXPORT", "NAME_CUSTOM", "RATIO"]]
    # display(max_scores_export)
    
    # Calculate best match for each name in custom dataset
    max_scores_custom = (cross_join[["NAME_EXPORT", "NAME_CUSTOM", "ratio"]]
            .groupby(by=["NAME_CUSTOM"])
            .apply(process_group)
            .rename(columns={"NAME_CUSTOM": "name_custom"})
            .reset_index()
            .drop(columns=["NAME_CUSTOM", "level_1"])
            .rename(columns={"name_custom": "NAME_CUSTOM", "ratio": "RATIO"}))
    max_scores_custom["STATE"] = state
    max_scores_custom = max_scores_custom[["STATE", "NAME_EXPORT", "NAME_CUSTOM", "RATIO"]]

    # Store matches in larger DataFrame
    top_matches_export = max_scores_export if top_matches_export is None \
        else pd.concat([top_matches_export, max_scores_export])
    top_matches_custom = max_scores_custom if top_matches_custom is None \
        else pd.concat([top_matches_custom, max_scores_custom])


12 record(s) in export, 13 in custom
14 record(s) in export, 14 in custom
3 record(s) in export, 2 in custom
2 record(s) in export, 5 in custom
7 record(s) in export, 11 in custom
0 record(s) in export, 0 in custom
Missing records in both dataframes for state "CO". Skipping.
2 record(s) in export, 1 in custom
0 record(s) in export, 0 in custom
Missing records in both dataframes for state "DC". Skipping.
2 record(s) in export, 1 in custom
5 record(s) in export, 1 in custom
15 record(s) in export, 16 in custom
0 record(s) in export, 0 in custom
Missing records in both dataframes for state "HI". Skipping.
23 record(s) in export, 20 in custom
0 record(s) in export, 0 in custom
Missing records in both dataframes for state "ID". Skipping.
5 record(s) in export, 5 in custom
29 record(s) in export, 29 in custom
5 record(s) in export, 4 in custom
18 record(s) in export, 14 in custom
5 record(s) in export, 6 in custom
2 record(s) in export, 2 in custom
2 record(s) in export, 2 in custom
4 record

In [41]:
print(len(top_matches_export))
print(len(top_matches_export["OBJECTID_EXPORT"].unique()))
top_matches_export["count"] = 1
dupes = top_matches_export.groupby(by="OBJECTID_EXPORT").sum().sort_values(by="count").query("count == 2").index.tolist()
top_matches_export.query("OBJECTID_EXPORT in @dupes")

452
448


OBJECTID_EXPORT STATE                      NAME_EXPORT  \
21               28    IA  KEOSAUQUA MUNICIPAL LIGHT & PWR   
22               28    IA  KEOSAUQUA MUNICIPAL LIGHT & PWR   
3              2011    MO                 CITY OF ROCKPORT   
4              2011    MO                 CITY OF ROCKPORT   
8              2151    OH               VILLAGE OF DESHLER   
9              2151    OH               VILLAGE OF DESHLER   
2              2484    PA             BOROUGH OF GOLDSBORO   
3              2484    PA             BOROUGH OF GOLDSBORO   

                      NAME_CUSTOM  RATIO  count  
21  ALTON MUNICIPAL LIGHT & POWER     68      1  
22        KEOSAUQUA LIGHT & POWER     68      1  
3               CITY OF ROCK PORT     47      1  
4              CITY OF ST. ROBERT     47      1  
8     DESHLER MUNICIPAL UTILITIES     58      1  
9     DESHLER MUNICIPAL UTILITIES     58      1  
2              BOROUGHT OF GIRARD     62      1  
3                GOLDSBORO - (PA)     62      1

In [42]:
top_matches_export.drop(columns="count").to_csv("exports2.csv", index=False)

In [47]:
manually_corrected = pd.read_csv("exports.csv")
manually_corrected = manually_corrected.query("IS_MATCH == 1")
manually_corrected.head()

OBJECTID_EXPORT STATE                    NAME_EXPORT  \
0             1116    AK  BIRCH CREEK VILLAGE ELEC UTIL   
1             1725    AK                 CITY OF AKUTAN   
4             2236    AK             CITY OF ELFIN COVE   
6              117    AK             CITY OF LARSEN BAY   
7             2726    AK    IPNATCHIAQ ELECTRIC COMPANY   

                     NAME_CUSTOM RATIO  IS_MATCH  
0   BIRCH CREEK ELECTRIC UTILITY   100         1  
1        AKUTAN ELECTRIC UTILITY   100         1  
4  ELFIN COVE UTILITY COMMISSION   100         1  
6             LARSEN BAY UTILITY   100         1  
7        IPNATCHIAQ ELECTRIC CO.   100         1

In [49]:
col_map = {"OBJECTID_EXPORT": "OBJECTID", "NAME_CUSTOM": "NAME"}
corrected3 = manually_corrected.rename(columns=col_map)[list(col_map.values())]
corrected3.head()

OBJECTID                           NAME
0      1116   BIRCH CREEK ELECTRIC UTILITY
1      1725        AKUTAN ELECTRIC UTILITY
4      2236  ELFIN COVE UTILITY COMMISSION
6       117             LARSEN BAY UTILITY
7      2726        IPNATCHIAQ ELECTRIC CO.

In [50]:
corrected = pd.concat([corrected, corrected3]).sort_values(by="OBJECTID")

In [52]:
corrected.drop_duplicates()

OBJECTID                                 NAME
0            1                      CITY OF AUGUSTA
1            7                      CITY OF AURELIA
2            8                       CITY OF KAPLAN
3            9                       CITY OF KASOTA
4           10                       CITY OF AURORA
...        ...                                  ...
1834      2943                     CITY OF SULLIVAN
1836      2945          SHAWANO MUNICIPAL UTILITIES
163       2946  COLDWATER BOARD OF PUBLIC UTILITIES
1838      2947       WAGONER PUBLIC WORKS AUTHORITY
402       2948             BLOOMER ELECTRIC UTILITY

[1761 rows x 2 columns]

In [58]:
merged = (municipalities
    .merge(right=corrected, how="left", on="OBJECTID")
    .loc[:, ["OBJECTID", "NAME_x", "NAME_y", "geometry"]])

merged["NAME"] = merged.apply(lambda r: r["NAME_y"] if isinstance(r["NAME_y"], str) else r["NAME_x"], axis=1)
merged[["OBJECTID", "NAME", "geometry"]]

OBJECTID                                 NAME  \
0            1                      CITY OF AUGUSTA   
1            7                      CITY OF AURELIA   
2            8                       CITY OF KAPLAN   
3            9                       CITY OF KASOTA   
4           10                       CITY OF AURORA   
...        ...                                  ...   
1840      2945          SHAWANO MUNICIPAL UTILITIES   
1841      2946  COLDWATER BOARD OF PUBLIC UTILITIES   
1842      2947       WAGONER PUBLIC WORKS AUTHORITY   
1843      2948             BLOOMER ELECTRIC UTILITY   
1844      2949       WALTERS PUBLIC WORKS AUTHORITY   

                                               geometry  
0     POLYGON ((-91.34713 35.43899, -91.34711 35.439...  
1     POLYGON ((-95.85936 42.90908, -95.85125 42.909...  
2     POLYGON ((-92.30731 30.00423, -92.30736 30.004...  
3     POLYGON ((-93.96939 44.30101, -93.96936 44.297...  
4     MULTIPOLYGON (((-103.38497 44.22756, -103.3849...  
...                                                 ...  
1840  MULTIPOLYGON (((-88.56882 44.79633, -88.56681 ...  
1841  MULTIPOLYGON (((-84.90152 41.94124, -84.88827 ...  
1842  MULTIPOLYGON (((-95.39223 35.97349, -95.38982 ...  
1843  MULTIPOLYGON (((-91.51448 45.09968, -91.50966 ...  
1844  MULTIPOLYGON (((-98.29907 34.36267, -98.29892 ...  

[1845 rows x 3 columns]

In [60]:
merged[["OBJECTID", "NAME"]].to_csv("corrected_municipal_utilities.csv")

In [61]:
merged["count"] = 1
merged[["OBJECTID", "count"]].groupby("OBJECTID").sum().sort_values("count")

count
OBJECTID       
1             1
2072          1
2070          1
2069          1
2065          1
...         ...
2949          1
2842          2
2885          2
2484          2
2151          2

[1841 rows x 1 columns]

In [72]:
merged.query("OBJECTID == 2885")
# LAFAYETTE, LA - LAFAYETTE UTILITIES SYSTEM, LAFAYETTE PUBLIC POWER AUTHORITY --> multiple city/state match
# WAYNE, NE - NORTHEAST POWER, WAYNE MUNICIPAL UTILITIES --> multiple city/state match
# BOROUGH OF GOLDSBORO, PA - BOROUGHT OF GIRARD, GOLDSBORO - (PA) --> how did that first result get there? Need to trace.
# VILLAGE OF DESHLER, OH - DESHLER MUNICIPAL UTILITIES --> duplicate in custom dataset

OBJECTID         NAME_x                     NAME_y  \
1779      2885  CITY OF WAYNE  WAYNE MUNICIPAL UTILITIES   
1780      2885  CITY OF WAYNE            NORTHEAST POWER   

                                               geometry  \
1779  POLYGON ((-97.03724 42.23611, -97.03608 42.235...   
1780  POLYGON ((-97.03724 42.23611, -97.03608 42.235...   

                           NAME  count  
1779  WAYNE MUNICIPAL UTILITIES      1  
1780            NORTHEAST POWER      1

In [73]:
municipalities[municipalities["NAME"].str.contains("CITY OF WAYNE")]

OBJECTID     ID                 NAME        ADDRESS   CITY STATE    ZIP  \
1523      1524  20224  CITY OF WAYNESVILLE           None   None    MO   None   
1525      1526  20226    CITY OF WAYNETOWN           None   None    IN   None   
2884      2885  20219        CITY OF WAYNE  306 PEARL ST.  WAYNE    NE  68787   

           TELEPHONE       TYPE COUNTRY NAICS_CODE  \
1523            None  MUNICIPAL     USA       2211   
1525            None  MUNICIPAL     USA       2211   
2884  (402) 375-2896  MUNICIPAL     USA       2211   

                                             NAICS_DESC  \
1523  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
1525  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
2884  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                                 SOURCE  SOURCEDATE  \
1523       EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06   
1525       EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06   
2884  EIA 860, EIA 861, Urban Areas - Cartographic B...  2021-10-06   

     VAL_METHOD    VAL_DATE                                        WEBSITE  \
1523      OTHER  2018-10-12                                           None   
1525      OTHER  2018-10-12                                           None   
2884      OTHER  2022-06-03  http://www.cityofwayne.org/index.aspx?NID=157   

     REGULATED                             CNTRL_AREA PLAN_AREA  \
1523      None  ASSOCIATED ELECTRIC COOPERATIVE, INC.      None   
1525      None               PJM INTERCONNECTION, LLC      None   
2884      None                   SOUTHWEST POWER POOL      None   

                      HOLDING_CO  SUMMR_PEAK  WINTR_PEAK  SUMMER_CAP  \
1523  CITY OF WAYNESVILLE - (MO)   -999999.0   -999999.0   -999999.0   
1525    CITY OF WAYNETOWN - (IN)   -999999.0   -999999.0   -999999.0   
2884               CITY OF WAYNE   -999999.0   -999999.0   -999999.0   

      WINTER_CAP  NET_GEN  PURCHASED  NET_EX  RETAIL_MWH  WSALE_MWH  \
1523   -999999.0    62690    -999999 -999999     -999999    -999999   
1525   -999999.0     5324    -999999 -999999     -999999    -999999   
2884   -999999.0    69183    -999999 -999999     -999999    -999999   

      TOTAL_MWH  TRANS_MWH  CUSTOMERS  YEAR  Shape__Are  Shape__Len  \
1523    -999999    -999999    -999999  2020    0.146293    1.545553   
1525    -999999    -999999        515  2020    0.138158    1.491906   
2884    -999999    -999999    -999999  2020    0.000508    0.170560   

                                               geometry  IS_EXPORT  
1523  POLYGON ((-92.40597 38.02059, -92.39108 38.020...          1  
1525  POLYGON ((-87.09252 40.21469, -87.08880 40.214...          1  
2884  POLYGON ((-97.03724 42.23611, -97.03608 42.235...          1

In [43]:
# import numpy as np
# combined["COUNT"] = 1
# combined = combined.replace({np.nan: 0, None: 0})
# combined[["IS_EXPORT", "IS_CUSTOM", "COUNT"]].groupby(["IS_EXPORT", "IS_CUSTOM"]).count()

COUNT
IS_EXPORT IS_CUSTOM       
0.0       1.0          823
1.0       0.0          716
          1.0         1125

#### Rural Cooperatives

In [44]:
coops_types = ['COOPERATIVE']
coops = gdf.query("TYPE in @coops_types")
coops.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Index: 812 entries, 3 to 2792
Data columns (total 37 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   OBJECTID    812 non-null    int64   
 1   ID          812 non-null    object  
 2   NAME        812 non-null    object  
 3   ADDRESS     812 non-null    object  
 4   CITY        812 non-null    object  
 5   STATE       812 non-null    object  
 6   ZIP         812 non-null    object  
 7   TELEPHONE   812 non-null    object  
 8   TYPE        812 non-null    object  
 9   COUNTRY     812 non-null    object  
 10  NAICS_CODE  812 non-null    object  
 11  NAICS_DESC  812 non-null    object  
 12  SOURCE      812 non-null    object  
 13  SOURCEDATE  812 non-null    object  
 14  VAL_METHOD  812 non-null    object  
 15  VAL_DATE    812 non-null    object  
 16  WEBSITE     812 non-null    object  
 17  REGULATED   812 non-null    object  
 18  CNTRL_AREA  812 non-null    object  
 19  PLAN

In [45]:
coops.head(2)

OBJECTID     ID                      NAME           ADDRESS         CITY  \
3         4  10009  KARNES ELECTRIC COOP INC  1007 N. HWY. 123  KARNES CITY   
4         5  10012         KAY ELECTRIC COOP     300 W. DOOLIN    BLACKWELL   

  STATE    ZIP       TELEPHONE         TYPE COUNTRY NAICS_CODE  \
3    TX  78118  (830) 780-3952  COOPERATIVE     USA       2211   
4    OK  74631  (580) 363-1260  COOPERATIVE     USA       2211   

                                          NAICS_DESC  \
3  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   
4  ELECTRIC POWER GENERATION, TRANSMISSION AND DI...   

                                              SOURCE  SOURCEDATE VAL_METHOD  \
3       EIA 861, TIGER/Line Shapefiles - U.S. Census  2021-10-06      OTHER   
4  EIA 861, TIGER/Line Shapefiles - U.S. Census, ...  2021-10-06      OTHER   

     VAL_DATE                   WEBSITE      REGULATED  \
3  2018-10-12  http://www.karnesec.org/  NOT AVAILABLE   
4  2019-05-01      www.kayelectric.coop  NOT AVAILABLE   

                                    CNTRL_AREA  \
3  ELECTRIC RELIABILITY COUNCIL OF TEXAS, INC.   
4                         SOUTHWEST POWER POOL   

                              PLAN_AREA                HOLDING_CO  SUMMR_PEAK  \
3                                 ERCOT  KARNES ELECTRIC COOP INC       207.0   
4  WESTERN FARMERS ELECTRIC COOPERATIVE         KAY ELECTRIC COOP        71.0   

   WINTR_PEAK  SUMMER_CAP  WINTER_CAP  NET_GEN  PURCHASED  NET_EX  RETAIL_MWH  \
3       208.0   -999999.0   -999999.0  -999999    1170464       0     1129899   
4        71.0   -999999.0   -999999.0        0     378230       0      371371   

   WSALE_MWH  TOTAL_MWH  TRANS_MWH  CUSTOMERS  YEAR  Shape__Are  Shape__Len  \
3    -999999    1170464          0      21327  2020    3.085930    8.976781   
4          0     378230          0       5785  2020    0.599957    3.978050   

                                            geometry  
3  POLYGON ((-98.54256 29.75188, -98.54102 29.751...  
4  POLYGON ((-97.74419 36.93999, -97.74638 36.977...